gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -277,13 +278,100 @@ impl GatewayLightning for ClnRpcService {
         Ok(tonic::Response::new(ReceiverStream::new(receiver)))
     }
 
-    type CompleteHtlcsStream = tonic::Streaming<CompleteHtlcsResponse>;
+    type CompleteHtlcsStream =
+        Pin<Box<dyn Stream<Item = Result<CompleteHtlcsResponse, tonic::Status>> + Send + 'static>>;
 
     async fn complete_htlcs(
         &self,
-        _request: tonic::Request<tonic::Streaming<CompleteHtlcsRequest>>,
+        request: tonic::Request<tonic::Streaming<CompleteHtlcsRequest>>,
     ) -> Result<tonic::Response<Self::CompleteHtlcsStream>, Status> {
-        unimplemented!()
+        let mut requests = request.into_inner();
+        let mut responses: Vec<Result<CompleteHtlcsResponse, tonic::Status>> = Vec::new();
+
+        while let Some(CompleteHtlcsRequest {
+            action,
+            intercepted_htlc_id,
+        }) = requests.message().await?
+        {
+            let hash = match sha256::Hash::from_slice(&intercepted_htlc_id) {
+                Ok(hash) => hash,
+                Err(e) => {
+                    error!("Invalid intercepted_htlc_id: {:?}", e);
+                    responses.push(Err(Status::invalid_argument(e.to_string())));
+                    continue;
+                }
+            };
+
+            // TODO: Implement non-blocking access to the outcomes map sowe can multithread complete_htlcs
+            if let Some(outcome) = self
+                .interceptor
+                .lock()
+                .await
+                .outcomes
+                .lock()
+                .await
+                .remove(&hash)
+            {
+                // Translate action request into a cln rpc response for `htlc_accepted` event
+                let htlca_res = match action {
+                    Some(Action::Settle(Settle { preimage })) => {
+                        if let Ok(pk) = XOnlyPublicKey::from_slice(&preimage) {
```

Comment:
It's very annoying we have to deal with all that parsing boilerplate here. Is there a way to get protobuf structs parsed into properly typed structs using some sort of derive macro?

Reply:
[pending] will look into easier type mapping between grpc and our rust types